infrastructure/src/main/java/org/corfudb/infrastructure/IServerRouter.java

```diff
@@ -170,4 +195,117 @@ default boolean messageIsValid(CorfuMsg msg, ChannelHandlerContext ctx) {
+     * @param ctx The context of the channel handler.
+     * @param clusterId The current cluster id.
+     */
+    default void sendWrongClusterError(HeaderMsg requestHeader, ChannelHandlerContext ctx, UuidMsg clusterId) {
+        HeaderMsg responseHeader = getHeaderMsg(requestHeader, false, true);
+        ResponseMsg response = getResponseMsg(responseHeader,
+                getWrongClusterErrorMsg(clusterId, requestHeader.getClusterId()));
+
+        sendResponse(response, ctx);
+
+        log.trace("sendWrongClusterError[{}]: Incoming request with a wrong cluster id, got {}, expected {}, " +
+                "request was: {}", requestHeader.getRequestId(), requestHeader.getClusterId(), clusterId, requestHeader);
+    }
+
+    /**
+     * Validate the epoch of an incoming request and send a WRONG_EPOCH error response
+     * if the server is in the wrong epoch.
+     * @param requestHeader The header of the incoming request to validate.
+     * @param ctx The context of the channel handler.
+     * @return True, if the epoch is correct, but false otherwise.
+     */
+    default boolean epochIsValid(HeaderMsg requestHeader, ChannelHandlerContext ctx) {
+        final long serverEpoch = getServerEpoch();
+        if(requestHeader.getEpoch() != serverEpoch) {
+            sendWrongEpochError(requestHeader, ctx);
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Validate that the cluster ID of the incoming request is equal to the cluster ID of a
+     * bootstrapped layout. If this is not the case, send a WRONG_CLUSTER error response.
+     * @param requestHeader The header of the incoming request to validate.
+     * @param ctx The context of the channel handler.
+     * @param layout The layout a server was bootstrapped with.
+     * @return True, if the cluster ID of the request matches the cluster ID of the layout, but false otherwise.
+     */
+    default boolean clusterIdIsValid(HeaderMsg requestHeader, ChannelHandlerContext ctx, Layout layout) {
+        UuidMsg currentClusterID = getUuidMsg(layout.getClusterId());
+        boolean match = requestHeader.getClusterId().equals(currentClusterID);
+
+        if(!match) {
+            sendWrongClusterError(requestHeader, ctx, currentClusterID);
+        }
+
+        return match;
+    }
+
+    /**
+     * Validate the incoming request. The request is valid if:
+     *    1) The flag ignoreEpoch is set to true, or it's set to false and the epoch is valid.
+     *    2) Also, if the flag ignoreClusterId is set to false,
+     *           a. The current layout server should be bootstrapped and
+     *           b. the request's cluster ID should be equal to the bootstrapped layout's cluster ID.
+     *
+     * @param req The incoming request.
+     * @param ctx The context of the channel handler.
+     * @return True if the request is valid, and false otherwise.
+     */
+    default boolean requestIsValid(RequestMsg req, ChannelHandlerContext ctx) {
```

Comment:
I am doing so with the methods above, but it's a bit awkward to do for `requestIsValid` since we need to distinguish between multiple error conditions. I can update the method name and Javadoc to specify that a response is sent in the case of an error.